File: StarBridge.Flutter/lib/features/account/legacy_profile_migration.dart
import 'account_port.dart';

/// Account-scoped migration port. User-entered credentials are transient request data only.
abstract interface class LegacyProfileMigrationPort {
  Future<LegacyProfileMigrationView> migrationStatus();
  Future<LegacyProfileMigrationView> previewMigration({
    LegacyAccountCredential? credential,
  });
  Future<LegacyProfileMigrationView> confirmMigration(
    String previewId, {
    required bool replaceExisting,
  });
}

final class LegacyProfileMigrationView {
  const LegacyProfileMigrationView(
    this.state, {
    this.previewId,
    this.sourceGameId,
    this.expiresAt,
    this.conflict = false,
    this.source,
    this.target,
  });
  final String state;
  final String? previewId;
  final String? sourceGameId;
  final DateTime? expiresAt;
  final bool conflict;
  final LegacyProfileMigrationSummary? source;
  final LegacyProfileMigrationSummary? target;

  factory LegacyProfileMigrationView.fromJson(Map<String, Object?> json) {
    final state = json['state'];
    if (state is! String || !_states.contains(state)) {
      throw const FormatException('Unknown migration state.');
    }
    final source = json['source'];
    final target = json['target'];
    final id = json['previewId'];
    final expires = DateTime.tryParse(json['expiresAt'] as String? ?? '');
    if (state == 'previewed' &&
        (id is! String ||
            id.isEmpty ||
            id.length > 64 ||
            source is! Map<String, Object?> ||
            target is! Map<String, Object?> ||
            expires == null)) {
      throw const FormatException('Incomplete migration preview.');
    }
    return LegacyProfileMigrationView(
      state,
      previewId: id as String?,
      sourceGameId: json['sourceGameId'] as String?,
      expiresAt: expires,
      conflict: json['conflict'] == true,
      source: source is Map<String, Object?>
          ? LegacyProfileMigrationSummary.fromJson(source)
          : null,
      target: target is Map<String, Object?>
          ? LegacyProfileMigrationSummary.fromJson(target)
          : null,
    );
  }
}

const _states = {
  'credentialRequired',
  'credentialRejected',
  'verificationRateLimited',
  'sourceNotConfigured',
  'profileNotFound',
  'notStarted',
  'previewed',
  'completed',
  'sourceUnavailable',
  'linkRequired',
  'accountMismatch',
  'authorizationRequired',
  'unsupportedData',
  'previewRequired',
  'previewExpired',
  'replacementRequired',
  'targetChanged',
};

final class LegacyProfileMigrationSummary {
  const LegacyProfileMigrationSummary({
    required this.callSign,
    required this.introduction,
    required this.modules,
    required this.favorites,
    required this.playTimeSeconds,
  });
  final String callSign;
  final String introduction;
  final int modules;
  final int favorites;
  final int? playTimeSeconds;
  factory LegacyProfileMigrationSummary.fromJson(Map<String, Object?> json) =>
      LegacyProfileMigrationSummary(
        callSign: json['callSign'] as String,
        introduction: json['introduction'] as String,
        modules: json['modules'] as int,
        favorites: json['favorites'] as int,
        playTimeSeconds: json['playTimeSeconds'] as int?,
      );
}

File: StarBridge.HostRuntime/Auth/LegacyProfileMigration.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Json;

namespace StarBridge.HostRuntime.Auth;

// Deliberately excludes legacy account ids, source keys, staged documents and all credentials.
public sealed record LegacyProfileMigrationSummary(
    string CallSign, string Introduction, int Modules, int Favorites, long? PlayTimeSeconds);
public sealed record LegacyProfileMigrationView(
    string State, string? PreviewId = null, string? SourceGameId = null,
    long SourceRevision = 0, long TargetRevision = 0, DateTimeOffset? ExpiresAt = null,
    bool Conflict = false, LegacyProfileMigrationSummary? Source = null,
    LegacyProfileMigrationSummary? Target = null, long? CompletedRevision = null);
public sealed record LegacyProfileMigrationResult(LegacyProfileMigrationView View, ScmOAuthSession ActiveSession);
public sealed record LegacyProfileCredential(string AccountName, string Password)
{
    public override string ToString() => "LegacyProfileCredential[REDACTED]";
}

public sealed partial class OAuthPkceClient
{
    public async Task<LegacyProfileMigrationResult> ExecuteLegacyProfileMigrationAsync(
        ScmOAuthSession session, string action, string? previewId, bool replaceExisting,
        CancellationToken cancellationToken, LegacyProfileCredential? credentials = null)
    {
        if (action is not ("status" or "preview" or "confirm"))
            throw new ArgumentException("Unknown migration operation.", nameof(action));
        if (action == "confirm" && !session.Capabilities.Contains("profile.write", StringComparer.Ordinal))
            throw new ScmApiForbiddenException("Migration requires profile.write.");
        var endpoint = new Uri(options.ResourceServerBaseUri,
            "app-api/starbridge/v1/legacy-profile-migration" + (action == "status" ? "" : "/" + action));
        async Task<LegacyProfileMigrationView> Send(ScmOAuthSession current, CancellationToken token)
            {
                using var request = new HttpRequestMessage(action == "status" ? HttpMethod.Get : HttpMethod.Post, endpoint);
                if (action == "confirm")
                    request.Content = JsonContent.Create(new { previewId, replaceExisting });
                else if (action == "preview" && credentials is not null)
                    request.Content = JsonContent.Create(new { credentials.AccountName, credentials.Password });
                using var response = await scmHttpClient.SendAsync(request, current.AccessToken, token,
                    action == "preview" ? TimeSpan.FromSeconds(45) : null);
                // Domain failures are safe allowlisted states, never arbitrary backend error strings.
                if (response.StatusCode is HttpStatusCode.Conflict or HttpStatusCode.ServiceUnavailable)
                {
                    var failed = await response.Content.ReadFromJsonAsync<LegacyProfileMigrationView>(token);
                    return new LegacyProfileMigrationView(AllowedFailure(failed?.State));
                }
                var result = await scmHttpClient.ReadJsonAsync<LegacyProfileMigrationView>(response, "Migration", token);
                if (result.State is not ("notStarted" or "previewed" or "completed") &&
                    AllowedFailure(result.State) != result.State)
                    throw new ScmApiProtocolException("Unknown migration response.");
                if (result.State == "previewed" &&
                    (string.IsNullOrWhiteSpace(result.PreviewId) || result.PreviewId.Length > 64 ||
                     result.Source is null || result.Target is null || result.ExpiresAt is null))
                    throw new ScmApiProtocolException("Incomplete migration preview.");
                return result;
            }
        // The short-lived write grant has no refresh token. Do not refresh it with the ordinary
        // read session, and do not replace the long-lived login session with this grant.
        if (action == "confirm")
            return new LegacyProfileMigrationResult(await Send(session, cancellationToken), session);
        var (view, activeSession) = await SendResourceRequestWithRefreshAndSessionAsync(
            session, Send, cancellationToken);
        _currentSession = activeSession;
        return new LegacyProfileMigrationResult(view, activeSession);
    }

    private static string AllowedFailure(string? code) => code switch
    {
        "linkRequired" or "accountMismatch" or "authorizationRequired" or "unsupportedData" or
        "previewRequired" or "previewExpired" or "replacementRequired" or "targetChanged" or
        "sourceNotConfigured" or "credentialRequired" or "credentialRejected" or
        "verificationRateLimited" or "profileNotFound" => code,
        _ => "sourceUnavailable"
    };
}
